### PR TITLE
Implement AWS KMS secret encryption backend

### PR DESCRIPTION
**Pull request by eseliger**
I've followed [the guide here](https://github.com/sourcegraph/sourcegraph/blob/main/internal/encryption/README.md) for implementing a new encryption backend and 
AWS doesn't seem to have a CRC32 checksum to test against, and there are no real "versions" of encryption keys(they can rotate in the background but as far as I understood it that's opaque to the API consumer), so I think the version function is not super useful here, but that's how far I got and understood the code. Can someone please check if this makes sense and tell me if something is missing here? 

Closes https://github.com/sourcegraph/sourcegraph/issues/19983

### CODE REVIEW

**Review by mrnugget — APPROVED** [review]
Changes look good to me. But I don't 100% understand this comment in the PR description:

> so I think the version function is not super useful here, but that's how far I got and understood the code.

Why is not useful? Because we don't support rotating keys in the database yet? That would apply to the env-var encryption backend too, right?

**Comment by eseliger** [comment]
Because for a given keyid the value never changes. Maybe I misunderstood how it's used though.

**Review by unknwon — APPROVED** [review]

**Comment by unknwon** [comment]
If you can wait until @arussellsaw's review, that would be the best, but if not, I don't think it should block merging and test with some real configurations before any customer is using the AWS KMS.

**Comment by eseliger** [comment]
Yes that was my plan :) There's no rush here since the next release is in over 3 weeks anyways.

File: internal/encryption/awskms/aws_kms_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/httptestutil"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+// testKeyID is the ID of a key defined here:
+// https://us-west-2.console.aws.amazon.com/kms/home?region=us-west-2#/kms/keys/4b739277-5a93-4551-b71c-99608c9c805d
+// If you want to update this test, feel free to replace the key ID used here.
+const testKeyID = "4b739277-5a93-4551-b71c-99608c9c805d"
+
+func TestRoundtrip(t *testing.T) {
```
Comment (arussellsaw) [inline internal/encryption/awskms/aws_kms_test.go:28]: considering this package is a very thin wrapper on top of the AWS library, how much value do we get out of a mocked test here? imo this kind of test doesn't really help your code safety, and just makes future changes more difficult, i'd be tempted to do something like i did with the google cloud implementation and have an integration behind a build tag

**Review by arussellsaw — APPROVED** [review]
if the key rotates opaquely behind the scenes i wouldn't worry too much about exposing key versions, the name of that method is a bit awkward, as it used to be called ID() but that didn't quite make sense either. we just want some way to concretely identify the key used to encrypt some data, if we were to hypothetically need to recover from some kind of issue, or to be able to check which key was used if we do a migration down the line.

File: internal/encryption/awskms/aws_kms_test.go (hunk 1 above)
Comment (eseliger) [reply internal/encryption/awskms/aws_kms_test.go:28]: It now does more things, like local encryption so I'm going to keep it in.

**Review by unknwon — APPROVED** [review]

**Comment by eseliger** [comment]
I talked to security about this and they weren't concerned with us merging this, and it will be target of a pen tester post merge, once we have it in. So I am going to merge this, given it works well locally.

**Pull request merged**